File: examples/scenarios/test-all-scenarios.js
/**
 * Test all scenarios for overlaps and missing people
 * Run this in the browser console on each scenario page
 */

async function testScenario(scenarioNumber) {
  const results = {
    scenario: `sc${String(scenarioNumber).padStart(2, '0')}`,
    timestamp: new Date().toISOString(),
    people: {
      expected: [],
      found: [],
      missing: [],
      duplicates: []
    },
    overlaps: {
      nodes: [],
      links: []
    },
    status: 'unknown'
  };

  try {
    // Get expected people from JSON data
    const sourceDataEl = document.querySelector('#sourceData');
    if (sourceDataEl) {
      const jsonText = sourceDataEl.textContent.trim();
      try {
        const jsonData = JSON.parse(jsonText);
        const treeData = Array.isArray(jsonData) ? jsonData : (jsonData.treeData || []);
        results.people.expected = treeData.map(p => ({
          id: p.id,
          name: p.data?.data?.full_name || p.data?.full_name || p.data?.data?.first_name || p.data?.first_name || p.id
        }));
      } catch (e) {
        console.error('Failed to parse JSON:', e);
      }
    }

    // Get found people from the tree
    const cards = Array.from(document.querySelectorAll('.f3 .card'));
    const foundNames = new Map();
    
    cards.forEach(card => {
      const labelEl = card.querySelector('.card-label');
      if (labelEl) {
        const name = labelEl.textContent.trim();
        const count = foundNames.get(name) || 0;
        foundNames.set(name, count + 1);
        
        results.people.found.push(name);
      }
    });

    // Check for duplicates
    foundNames.forEach((count, name) => {
      if (count > 1) {
        results.people.duplicates.push({ name, count });
      }
    });

    // Check for missing people
    const foundNamesSet = new Set(results.people.found);
    results.people.expected.forEach(expected => {
      const found = foundNamesSet.has(expected.name);
      if (!found) {
        results.people.missing.push(expected);
      }
    });

    // Check for overlapping nodes (cards with same position)
    const cardPositions = new Map();
    cards.forEach(card => {
      const rect = card.getBoundingClientRect();
      const key = `${Math.round(rect.left)},${Math.round(rect.top)}`;
      if (cardPositions.has(key)) {
        const existing = cardPositions.get(key);
        results.overlaps.nodes.push({
          name1: existing.name,
          name2: card.querySelector('.card-label')?.textContent.trim() || 'unknown',
          position: key
        });
      } else {
        const name = card.querySelector('.card-label')?.textContent.trim() || 'unknown';
        cardPositions.set(key, { name, card });
      }
    });

    // Check for overlapping links (lines that cross or overlap)
    const links = Array.from(document.querySelectorAll('.f3 .link'));
    const linkPositions = [];
    links.forEach(link => {
      const path = link.querySelector('path');
      if (path) {
        const d = path.getAttribute('d');
        const rect = link.getBoundingClientRect();
        linkPositions.push({
          d,
          rect: {
            x: rect.left,
            y: rect.top,
            width: rect.width,
            height: rect.height
          }
        });
      }
    });

    // Simple overlap check: links with overlapping bounding boxes
    for (let i = 0; i < linkPositions.length; i++) {
      for (let j = i + 1; j < linkPositions.length; j++) {
        const link1 = linkPositions[i];
        const link2 = linkPositions[j];
        
        // Check if bounding boxes overlap significantly
        const overlapX = Math.max(0, Math.min(link1.rect.x + link1.rect.width, link2.rect.x + link2.rect.width) - Math.max(link1.rect.x, link2.rect.x));
        const overlapY = Math.max(0, Math.min(link1.rect.y + link1.rect.height, link2.rect.y + link2.rect.height) - Math.max(link1.rect.y, link2.rect.y));
        const overlapArea = overlapX * overlapY;
        const minArea = Math.min(link1.rect.width * link1.rect.height, link2.rect.width * link2.rect.height);
        
        // If overlap is more than 50% of the smaller link, consider it an overlap
        if (overlapArea > minArea * 0.5 && overlapArea > 100) {
          results.overlaps.links.push({
            link1: i,
            link2: j,
            overlapArea: Math.round(overlapArea)
          });
        }
      }
    }

    // Determine overall status
    const hasIssues = results.people.missing.length > 0 || 
                     results.people.duplicates.length > 0 ||
                     results.overlaps.nodes.length > 0 ||
                     results.overlaps.links.length > 0;

    results.status = hasIssues ? 'FAILED' : 'PASSED';
    results.summary = {
      totalExpected: results.people.expected.length,
      totalFound: results.people.found.length,
      missingCount: results.people.missing.length,
      duplicateCount: results.people.duplicates.length,
      nodeOverlapCount: results.overlaps.nodes.length,
      linkOverlapCount: results.overlaps.links.length
    };

  } catch (error) {
    results.status = 'ERROR';
    results.error = error.message;
    console.error('Error testing scenario:', error);
  }

  return results;
}

// Export for use
if (typeof window !== 'undefined') {
  window.testScenario = testScenario;
}

// Auto-run if on a scenario page
if (window.location.pathname.includes('sc') && window.location.pathname.includes('.html')) {
  const match = window.location.pathname.match(/sc(\d+)\.html/);
  if (match) {
    const scenarioNum = parseInt(match[1], 10);
    testScenario(scenarioNum).then(results => {
      console.log(`\n=== ${results.scenario.toUpperCase()} TEST RESULTS ===`);
      console.log('Status:', results.status);
      console.log('Summary:', results.summary);
      if (results.people.missing.length > 0) {
        console.warn('Missing people:', results.people.missing);
      }
      if (results.people.duplicates.length > 0) {
        console.warn('Duplicate people:', results.people.duplicates);
      }
      if (results.overlaps.nodes.length > 0) {
        console.warn('Overlapping nodes:', results.overlaps.nodes);
      }
      if (results.overlaps.links.length > 0) {
        console.warn('Overlapping links:', results.overlaps.links);
      }
      console.log('Full results:', results);
    });
  }
}

File: examples/scenarios/test-all-scenarios.py
#!/usr/bin/env python3
"""
Test all scenarios sc01-sc25 for overlaps and missing people.
This script reads the JSON files and checks the rendered HTML.
"""

import json
import os
import sys
from pathlib import Path

def get_expected_people(json_file):
    """Extract expected people from JSON file."""
    try:
        with open(json_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Handle both array and object formats
        if isinstance(data, list):
            tree_data = data
        elif isinstance(data, dict) and 'treeData' in data:
            tree_data = data['treeData']
        else:
            tree_data = []
        
        expected = []
        for person in tree_data:
            person_id = person.get('id', '')
            person_data = person.get('data', {})
            
            # Handle nested data.data structure
            if isinstance(person_data, dict) and 'data' in person_data:
                inner_data = person_data['data']
            else:
                inner_data = person_data
            
            # Get name from various possible fields
            full_name = (
                inner_data.get('full_name') or
                inner_data.get('first_name') or
                inner_data.get('preferred_name') or
                person_id
            )
            
            # If we have first_name and last_name, combine them
            if not full_name or full_name == person_id:
                first_name = inner_data.get('first_name', '')
                last_name = inner_data.get('last_name', '')
                if first_name and last_name:
                    full_name = f"{first_name} {last_name}"
                elif first_name:
                    full_name = first_name
                elif last_name:
                    full_name = last_name
            
            expected.append({
                'id': person_id,
                'name': full_name or person_id
            })
        
        return expected
    except Exception as e:
        print(f"Error reading {json_file}: {e}")
        return []

def test_scenario(scenario_num):
    """Test a single scenario."""
    scenario_id = f"sc{scenario_num:02d}"
    json_file = Path(__file__).parent / f"{scenario_id}.json"
    
    if not json_file.exists():
        return {
            'scenario': scenario_id,
            'status': 'ERROR',
            'error': f'JSON file not found: {json_file}'
        }
    
    expected_people = get_expected_people(json_file)
    
    return {
        'scenario': scenario_id,
        'expected_count': len(expected_people),
        'expected_people': [p['name'] for p in expected_people],
        'json_file': str(json_file),
        'html_file': f"{scenario_id}.html"
    }

def main():
    """Test all scenarios sc01-sc25."""
    results = []
    
    print("Testing all scenarios sc01-sc25...")
    print("=" * 60)
    
    for i in range(1, 26):
        result = test_scenario(i)
        results.append(result)
        
        status_icon = "✓" if result.get('status') != 'ERROR' else "✗"
        expected_count = result.get('expected_count', 0)
        print(f"{status_icon} {result['scenario']}: {expected_count} expected people")
        
        if result.get('status') == 'ERROR':
            print(f"    ERROR: {result.get('error', 'Unknown error')}")
    
    print("=" * 60)
    print(f"\nSummary:")
    print(f"  Total scenarios: {len(results)}")
    print(f"  Successful: {sum(1 for r in results if r.get('status') != 'ERROR')}")
    print(f"  Errors: {sum(1 for r in results if r.get('status') == 'ERROR')}")
    
    # Save results to file
    output_file = Path(__file__).parent / 'test-results.json'
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(results, f, indent=2)
    
    print(f"\nResults saved to: {output_file}")
    print("\nNote: This script only validates JSON files.")
    print("      To check for overlaps and missing people in rendered HTML,")
    print("      use the browser test function on each scenario page.")

if __name__ == '__main__':
    main()

